In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [ ]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
applicationDropDF = application_df.drop(['EIN', 'NAME', 'AFFILIATION'], axis=1)
applicationDropDF

APPLICATION_TYPE CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0                  T10          C1000    ProductDev   Association       1   
1                   T3          C2000  Preservation  Co-operative       1   
2                   T5          C3000    ProductDev   Association       1   
3                   T3          C2000  Preservation         Trust       1   
4                   T3          C1000     Heathcare         Trust       1   
...                ...            ...           ...           ...     ...   
34294               T4          C1000    ProductDev   Association       1   
34295               T4          C3000    ProductDev   Association       1   
34296               T3          C2000  Preservation   Association       1   
34297               T5          C3000    ProductDev   Association       1   
34298               T3          C1000  Preservation  Co-operative       1   

          INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  IS_SUCCESSFUL  
0                  0                      N      5000              1  
1             1-9999                      N    108590              1  
2                  0                      N      5000              0  
3        10000-24999                      N      6692              1  
4      100000-499999                      N    142590              1  
...              ...                    ...       ...            ...  
34294              0                      N      5000              0  
34295              0                      N      5000              0  
34296              0                      N      5000              0  
34297              0                      N      5000              1  
34298          1M-5M                      N  36500179              0  

[34299 rows x 9 columns]

In [ ]:
# Determine the number of unique values in each column.
n = applicationDropDF.nunique(axis=0)
  
print("No.of.unique values in each column :\n",
      n)

No.of.unique values in each column :
 APPLICATION_TYPE            17
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [ ]:
# Look at APPLICATION_TYPE value counts for binning
applicationDropDF.APPLICATION_TYPE.value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
# Replace in dataframe
for app in applicationDropDF:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Look at CLASSIFICATION value counts for binning
applicationDropDF.CLASSIFICATION.value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [ ]:
# You may find it helpful to look at CLASSIFICATION value counts >1
applicationDropDF.groupby('CLASSIFICATION').filter(lambda x: len(x) >= 2).CLASSIFICATION.value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C4100        6
C1720        6
C2400        6
C1600        5
C1257        5
C1260        3
C2710        3
C0           3
C1267        2
C1256        2
C1234        2
C1246        2
C3200        2
Name: CLASSIFICATION, dtype: int64

In [ ]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
# Define cutoff value
cutoff = 10

# Create a list of classifications with count less than the cutoff
class_counts = application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = list(class_counts[class_counts < cutoff].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Other       98
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
Name: CLASSIFICATION, dtype: int64

In [ ]:
# Convert categorical data to numeric with `pd.get_dummies`
with_dummies_df = pd.get_dummies(applicationDropDF)
with_dummies_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0           1      5000              1                     1   
1           1    108590              1                     0   
2           1      5000              0                     0   
3           1      6692              1                     0   
4           1    142590              1                     0   
...       ...       ...            ...                   ...   
34294       1      5000              0                     0   
34295       1      5000              0                     0   
34296       1      5000              0                     0   
34297       1      5000              1                     0   
34298       1  36500179              0                     0   

       APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  \
0                         0                     0                     0   
1                         0                     0                     0   
2                         0                     0                     0   
3                         0                     0                     0   
4                         0                     0                     0   
...                     ...                   ...                   ...   
34294                     0                     0                     0   
34295                     0                     0                     0   
34296                     0                     0                     0   
34297                     0                     0                     0   
34298                     0                     0                     0   

       APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  ...  \
0                         0                     0                     0  ...   
1                         0                     0                     0  ...   
2                         0                     0                     0  ...   
3                         0                     0                     0  ...   
4                         0                     0                     0  ...   
...                     ...                   ...                   ...  ...   
34294                     0                     0                     0  ...   
34295                     0                     0                     0  ...   
34296                     0                     0                     0  ...   
34297                     0                     0                     0  ...   
34298                     0                     0                     0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                      0                       0                         0   
1                      1                       0                         0   
2                      0                       0                         0   
3                      0                       1                         0   
4                      0                       0                         1   
...                  ...                     ...                       ...   
34294                  0                       0                         0   
34295                  0                       0                         0   
34296                  0                       0                         0   
34297                  0                       0                         0   
34298                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                       0                 0                       0   
1                       0                 0                       0   
2                       0                 0                       0   
3                       0                 0                       0   
4                       0                 0                       0   
...          

In [ ]:
# Split our preprocessed data into our features and target arrays
y = with_dummies_df["IS_SUCCESSFUL"].values
X = with_dummies_df.drop(["IS_SUCCESSFUL"],1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)


<ipython-input-10-f80a2c389a95>:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  X = with_dummies_df.drop(["IS_SUCCESSFUL"],1).values


In [ ]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [ ]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 = 140
hidden_nodes_layer2 = 50

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 140)               15540     
                                                                 
 dense_1 (Dense)             (None, 50)                7050      
                                                                 
 dense_2 (Dense)             (None, 1)                 51        
                                                                 
Total params: 22,641
Trainable params: 22,641
Non-trainable params: 0
_________________________________________________________________


In [ ]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"]) 

In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=120) 

Epoch 1/120
804/804 [==============================] - 3s 2ms/step - loss: 0.6275 - accuracy: 0.6446
Epoch 2/120
804/804 [==============================] - 2s 2ms/step - loss: 0.6112 - accuracy: 0.6541
Epoch 3/120
804/804 [==============================] - 2s 2ms/step - loss: 0.6080 - accuracy: 0.6565
Epoch 4/120
804/804 [==============================] - 2s 3ms/step - loss: 0.6067 - accuracy: 0.6564
Epoch 5/120
804/804 [==============================] - 2s 2ms/step - loss: 0.6052 - accuracy: 0.6578
Epoch 6/120
804/804 [==============================] - 2s 3ms/step - loss: 0.6041 - accuracy: 0.6596
Epoch 7/120
804/804 [==============================] - 3s 3ms/step - loss: 0.6033 - accuracy: 0.6579
Epoch 8/120
804/804 [==============================] - 2s 2ms/step - loss: 0.6025 - accuracy: 0.6593
Epoch 9/120
804/804 [==============================] - 2s 3ms/step - loss: 0.6014 - accuracy: 0.6595
Epoch 10/120
804/804 [==============================] - 2s 2ms/step - loss: 0.6016 - accura

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.6202 - accuracy: 0.6562 - 519ms/epoch - 2ms/step
Loss: 0.6202421188354492, Accuracy: 0.6562098860740662


In [ ]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity_Optimization.h5')

# **Predicting Successful Applicants Who Receive Funding From Alphabet Soup**

Using neural networks, I have developed a tool that can help identify which applicants who seek funding have the best chance of success in their ventures. 


**Defining Target and Feature Variables**

The target variable is the model_accuracy. The feature variables include the columns, bins, epochs, neurons, and layers. The EIN and NAME were removed from the analysis since they are identification columns and are neither targets nor features.


**Optimization Methods**

In my first attempt at optimizing the model, I dropped the AFFILIATION column and increased hidden nodes layer1 to from 80 to 120, and hidden nodes layer2 from 30 to 45. This had a negligible impact, however. Since adding more layers does not guarantee more accurate results, I was curious to experiment with reducing the number of layers, and reduced hidden nodes layer1 to 40, and hidden nodes layer2 to 20. This unfortunately reduced accuracy by 9%, from 73% to 66% . Finally, I increased hidden nodes layer1 to 140, and hidden nodes layer2 to 50, and also increased the number of epochs from 100 to 120. This also had a 66% prediction accuracy.


**Ideas For Better Optimization**

Some ideas for improving predictions in future optimization attempts include dropping more columns, creating more bins for rare occurrences in columns, and increasing or decreasing the number of values for each bin.
